tactic/monotonicity/basic.lean

```diff
@@ -0,0 +1,163 @@
+
+import tactic.basic
+import category.basic
+import algebra.order_functions
+import algebra.order
+import meta.rb_map
+
+namespace tactic.interactive
+open tactic list
+open lean lean.parser  interactive
+open interactive.types
+
+structure mono_cfg :=
+  (unify := ff)
+
+@[derive [decidable_eq,has_reflect]]
+inductive mono_selection : Type
+| left : mono_selection
+| right : mono_selection
+| both : mono_selection
+
+section compare
+
+parameter opt : mono_cfg
+
+meta def compare (e₀ e₁ : expr) : tactic unit := do
+if opt.unify then do
+guard (¬ e₀.is_mvar ∧ ¬ e₁.is_mvar),
+unify e₀ e₁
+else is_def_eq e₀ e₁
+
+meta def find_one_difference
+: list expr → list expr → tactic (list expr × expr × expr × list expr)
+ | (x :: xs) (y :: ys) :=
+   do c ← try_core (compare x y),
+      if c.is_some
+      then prod.map (cons x) id <$> find_one_difference xs ys
+      else do
+        guard (xs.length = ys.length),
+        mzip_with' compare xs ys,
+        return ([],x,y,xs)
+ | xs ys := fail format!"find_one_difference: {xs}, {ys}"
+
+end compare
+
+def last_two {α : Type*} (l : list α) : option (α × α) :=
+match l.reverse with
+| (x₁ :: x₀ :: _) := some (x₀, x₁)
+| _ := none
+end
+
+meta def match_imp : expr → tactic (expr × expr)
+ | `(%%e₀ → %%e₁) :=
+   do guard (¬ e₁.has_var),
+      return (e₀,e₁)
+ | _ := failed
+
+open expr
+
+meta def same_operator : expr → expr → bool
+| (app e₀ _) (app e₁ _) :=
+let fn₀ := e₀.get_app_fn,
+    fn₁ := e₁.get_app_fn in
+    fn₀.is_constant ∧ fn₀.const_name = fn₁.const_name
+| (pi _ _ _ _) (pi _ _ _ _) := tt
+| _ _ := ff
+
+meta def get_operator (e : expr) : option name :=
+guard (¬ e.is_pi) >> pure e.get_app_fn.const_name
+
+meta def monotoncity.check_rel (xs : list expr) (l r : expr) : tactic (option name) :=
+do guard (same_operator l r) <|>
+     do { fail format!"{l} and {r} should be the f x and f y for some f" },
+   if l.is_pi then pure none
+              else pure r.get_app_fn.const_name
+
+@[reducible]
+def mono_key := (with_bot name × with_bot name)
```

Comment:
Can't you use `name` for this? I believe `name.anonymous` functions as a bottom element for the type

Reply:
That's true, I just want to make sure to differentiate between the case where we have a pi construct and a function.